runtime/queries/metricsview.go

```diff
@@ -578,3 +579,20 @@ func (q *MetricsViewRows) generateFilename(mv *runtimev1.MetricsView) string {
 	}
 	return filename
 }
+
+func resolveMVAndPolicy(ctx context.Context, rt *runtime.Runtime, instanceID, metricsViewName string) (*runtimev1.MetricsView, *runtime.ResolvedMetricsViewPolicy, error) {
+	mv, lastUpdatedOn, err := lookupMetricsView(ctx, rt, instanceID, metricsViewName)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	resolvedPolicy, err := rt.ResolveMetricsViewPolicy(auth.GetClaims(ctx).Attributes(), instanceID, mv, lastUpdatedOn)
+	if err != nil {
+		return nil, nil, err
+	}
+	if resolvedPolicy != nil && !resolvedPolicy.HasAccess {
+		return nil, nil, status.Error(codes.Unauthenticated, "action not allowed")
+	}
```

Comment:
What about:

- Computing the resolved policy in the server API handler
- Then setting it as a property on the query before resolving it

That should make it available for use in `Key()`, avoiding the hacks around casting to `MetricViewQuery`. It would also avoid mixing authentication context (which is scoped to `runtime/server`) into the `runtime/queries` package (which may be called from outside server scope, such is in the upcoming scheduled reports feature).

Is that possible or does anything require the policy to be resolved inside the `queries` package?